### ISSUE TITLE
FreeBSD support

### ISSUE THREAD

**Issue by Ornias1993** [unifi-poller/unifi-poller#170]
FreeBSD support is very much possible.
Although I didn't manage to get the manpages to work, building the app is very much possible:

short how-to:
- Make sure "Bash" and "gmake" are installed (additional dependencies)
- git clone -b v1.6.3 https://github.com/unifi-poller/unifi-poller.git
- cd unifi-poller
- dep ensure
- gmake

Considering Unifi Controller is also available within the pkg repository of freebsd and the minimal amount of work required to get this to work, I think it's worthwhile adding support to both the documentation and future builds.

Still TODO:
- Getting manpages to work
- making a nicer rc.d file (I did make a quick hack though)
- submitting it to the freebsd package repository (pkg install unifi-poller would be awesome)

**Comment by davidnewhall** [comment]
> Considering Unifi Controller is also available within the pkg repository of freebsd

Wait, what? It's already been packaged for FreeBSD? Please elaborate on this statement; maybe I'm misunderstanding. If it has been packaged, please point me to that package, and any build system that made it?

I'm very much down to add support for more operating systems, especially if people are able to help out and contribute. You seem willing and able, so lets work this out! I'm nearly ready to release v2, so adding FreeBSD support is best done there, or shortly after in a v2.0.1 release. I'm already building this for darwin (macOS) and adding it to a homebrew repo, but that's probably not even close to what FreeBSD requires. That said, cross compiling isn't likely the hurdle I'm going to run into.

The short end of the story is that I don't know much about FreeBSD. I've never installed it, I've made a BSD package and I've never managed a package repo for it; I'm just lost. I've logged into and had root on FreeBSD systems, but I wouldn't consider myself proficient beyond "I know mac and linux very well and they're similar."

I guess to start with, I'll ask a few questions and maybe in the next week or two we can chat on Discord? Next questions: Can you point me to docs, or provide a general explanation of the freebsd package building process? Nextly, I'll need to know how package repos work on FreeBSD, so if you have docs or pointers for that, please provide them! I currently host the linux repos on Jfrog Bintray, but I don't see anything that says it supports a BSD repo. Maybe it does, and it's just a generic source? Not sure yet.

I realize I can probably google much of this, and I'll probably install a freebsd VM soon to mess around, but your knowledge may speed up the integration process. I'm currently at the hospital for the weekend because my son got sick (he's going through chemotherapy and his condition sucks up a lot of my time). I'll be more available to discuss and look into this further in a few days to a week or two. Just not sure how things are going to turn out right this moment. New year, new things to worry about. :)

Thanks for opening this issue!

**Comment by Ornias1993** [comment]
We'll I myself also never build any packages, I just managed to get it to build and run...
Which should be most of the work...

Anyway about the UnifiController:
It has been ported some time ago and the port has been pretty well maintained:
`pkg install unifi5` should do the trick for installing it (taking into account some dependancies)

see also:
https://www.freshports.org/net-mgmt/unifi5/

**Comment by davidnewhall** [comment]
ahh. I gotcha.. the controller was ported. I misunderstood. I'll have to do more research on building FreeBSD packages and how to host them. Will take some time..

**Comment by toomanyonhere** [comment]
Just to let you know, the git master branch builds and runs fine on my FreeBSD-Current (13) box.  I just did this:
 % go get github.com/unifi-poller/unifi-poller
since the make file uses gmake-isms go build <...> didn't work. but cd-ing into ~/go/src/github/unifi-poller/unifi-poller  and running 
% gmake install
it built fine but error-ed out that direct installing was only configured for macos. It built the binary which I manually moved to /usr/local/bin and then copied the up.yaml.example file into /etc/unifi-poller (which is the wrong place for FreeBSD it should be /usr/local/etc... but I was just trying to see if this would work.)  and edit the file for my controller and set it prometheus.   For the first run I had debug turned on.  
I haven't made a service file for it yet to run it a daemon on startup as I was just trying to see if it would run on FreeBSD, All the other exporters written in go that I have played with  have compiled and run fine.  I usually run my unifi exporter on the virtual machine running the controller butthat exporter died so I started looking for a new one and found yours.  My prometheus is running on my FreeBSD box.

**Comment by davidnewhall** [comment]
This is great info, thank you for sharing! When you make that service file, can you share it please? I'll get some time eventually to come back and work on this project. I'd love to add FreeBSD support and create a working package. Are you on Discord? Maybe I can hit you up to do some testing later? (later, in like a couple weeks or so, I'm very tied up atm) Thanks!!

**Comment by toomanyonhere** [comment]
Sure, I'll help with what I can, programming is not my strong suit but I can do testing and verify things are working.  I'm not presently on discord I signed up for it then deleted the app.  I'll try to get the service file written soon, they are fairly simple for this use case. I need to write on for a different exporter, too.

**Comment by davidnewhall** [comment]
I've added package building support for FreeBSD and included a service file to allow it to start.
Would someone mind giving the attached package a shot? (unzip it first, github made me zip it)
```shell
pkg install unifi-poller-2.0.0-alfalfa1_719.amd64.txz
echo unifipoller_enable="YES" >> /etc/rc.conf
service unifi-poller start
```
[unifi-poller-2.0.0-alfalfa1_719.amd64.txz.zip](https://github.com/unifi-poller/unifi-poller/files/4142337/unifi-poller-2.0.0-alfalfa1_719.amd64.txz.zip)

The config file is at `/usr/local/etc/unifi-poller/up.conf` and the man file should also be installed, check with `man unifi-poller`. Thanks for the pointers!

**Comment by toomanyonhere** [comment]
Nice!  sorry I didn't have time to write the service file.  The pkg installed fine in my machine and after copying my working config file to /usr/local/etc/unifi-poller and started it as a service it is running fine!
Great job and thank-you!

**Comment by davidnewhall** [comment]
Great, glad it's working! I don't know anything about FreeBSD repos, but I'll look into hosting one somehow. Going forward, at minimum, the new package will be available as of the 2.0 release on the github releases page.
